File: inference 2.py
import numpy as np
import tensorflow as tf

def load_tflite_model(model_path):
    interpreter = tf.lite.Interpreter(model_path=model_path)
    interpreter.allocate_tensors()
    return interpreter

def preprocess_input(input_data, expected_shape):
    # Resize if needed (for safety)
    if input_data.shape != tuple(expected_shape[1:]):
        input_data = tf.image.resize(input_data, expected_shape[1:3]).numpy()
    
    # Normalize input data to [0,1] range or as needed
    input_data = input_data.astype(np.float32) / 255.0

    # Expand dims if needed
    if len(input_data.shape) == len(expected_shape) - 1:
        input_data = np.expand_dims(input_data, axis=0)

    return input_data

def run_tflite_inference(interpreter, input_data):
    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()
    
    input_shape = input_details[0]['shape']
    input_data = preprocess_input(input_data, input_shape)

    interpreter.set_tensor(input_details[0]['index'], input_data)
    interpreter.invoke()
    output = interpreter.get_tensor(output_details[0]['index'])
    return output

# --- Load interpreters for each stream ---
rgb_model_path = "tflite_models/best_models/best_rgb_feature_model_quantized.tflite"
motion_model_path = "tflite_models/best_models/best_motion_feature_model_quantized.tflite"
lstm_model_path = "tflite_models/best_models/best_combined_lstm_classifier_quantized.tflite"

rgb_interpreter = load_tflite_model(rgb_model_path)
motion_interpreter = load_tflite_model(motion_model_path)
lstm_interpreter = load_tflite_model(lstm_model_path)

# --- Simulated Inputs (replace with real video frames/slices) ---
# Assume 10 time steps (frames) per window, 224x224 resolution, 3 channels (RGB)
sequence_length = 10
frame_height, frame_width, channels = 224, 224, 3

rgb_sequence = np.random.rand(sequence_length, frame_height, frame_width, channels)
motion_sequence = np.random.rand(sequence_length, frame_height, frame_width, 1)  # grayscale or optical flow

# --- Extract features per frame (RGB & motion streams) ---
rgb_features = []
motion_features = []

for i in range(sequence_length):
    rgb_feat = run_tflite_inference(rgb_interpreter, rgb_sequence[i])
    motion_feat = run_tflite_inference(motion_interpreter, motion_sequence[i])
    rgb_features.append(rgb_feat)
    motion_features.append(motion_feat)

rgb_features = np.array(rgb_features)
motion_features = np.array(motion_features)

# --- Combine features ---
# You might concatenate or fuse features based on your training architecture
# Example: concatenate along feature dimension
combined_features = np.concatenate([rgb_features, motion_features], axis=-1)
combined_features = np.expand_dims(combined_features, axis=0)  # shape: (1, seq_len, feature_dim)

# --- Run final LSTM + Dense inference ---
input_details = lstm_interpreter.get_input_details()
output_details = lstm_interpreter.get_output_details()

# Handle quantization if needed
if input_details[0]['dtype'] == np.uint8:
    scale, zero_point = input_details[0]['quantization']
    combined_features = combined_features / scale + zero_point
    combined_features = combined_features.astype(np.uint8)

lstm_interpreter.set_tensor(input_details[0]['index'], combined_features)
lstm_interpreter.invoke()
final_output = lstm_interpreter.get_tensor(output_details[0]['index'])

# --- Print prediction ---
print("Final Prediction:", final_output)

File: inference 2.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <tensorflow/lite/interpreter.h>
#include <tensorflow/lite/kernels/register.h>
#include <tensorflow/lite/model.h>

using namespace tflite;

std::unique_ptr<tflite::Interpreter> LoadModel(const std::string& model_path) {
    auto model = tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
    if (!model) {
        std::cerr << "Failed to mmap model " << model_path << std::endl;
        exit(-1);
    }

    ops::builtin::BuiltinOpResolver resolver;
    std::unique_ptr<Interpreter> interpreter;
    InterpreterBuilder(*model, resolver)(&interpreter);
    if (!interpreter) {
        std::cerr << "Failed to construct interpreter for " << model_path << std::endl;
        exit(-1);
    }

    if (interpreter->AllocateTensors() != kTfLiteOk) {
        std::cerr << "Failed to allocate tensors!" << std::endl;
        exit(-1);
    }

    return interpreter;
}

void FillInputTensor(tflite::Interpreter* interpreter, const std::vector<float>& data, int input_idx) {
    float* input = interpreter->typed_input_tensor<float>(input_idx);
    std::copy(data.begin(), data.end(), input);
}

std::vector<float> RunInference(tflite::Interpreter* interpreter) {
    if (interpreter->Invoke() != kTfLiteOk) {
        std::cerr << "Error during inference." << std::endl;
        exit(-1);
    }

    float* output = interpreter->typed_output_tensor<float>(0);
    int output_size = interpreter->tensor(interpreter->outputs()[0])->bytes / sizeof(float);
    return std::vector<float>(output, output + output_size);
}

int main() {
    // Load models
    auto rgb_interpreter = LoadModel("models/rgb_model.tflite");
    auto motion_interpreter = LoadModel("models/motion_model.tflite");
    auto lstm_interpreter = LoadModel("models/lstm_model.tflite");

    // Simulated input (replace with actual data from preprocessing)
    std::vector<float> rgb_input(30 * 7 * 7 * 1280, 0.5f);      // e.g., (30,7,7,1280) flattened
    std::vector<float> motion_input(30 * 7 * 7 * 64, 0.5f);     // e.g., (30,7,7,64) flattened

    // Step 1: Run RGB stream
    FillInputTensor(rgb_interpreter.get(), rgb_input, 0);
    auto rgb_output = RunInference(rgb_interpreter.get());  // e.g., shape (30, 128)

    // Step 2: Run Motion stream
    FillInputTensor(motion_interpreter.get(), motion_input, 0);
    auto motion_output = RunInference(motion_interpreter.get());  // e.g., shape (30, 64)

    // Step 3: Concatenate features for LSTM input
    std::vector<float> lstm_input;
    for (int i = 0; i < 30; ++i) {
        lstm_input.insert(lstm_input.end(), 
                          rgb_output.begin() + i * 128, 
                          rgb_output.begin() + (i + 1) * 128);
        lstm_input.insert(lstm_input.end(), 
                          motion_output.begin() + i * 64, 
                          motion_output.begin() + (i + 1) * 64);
    }

    // Step 4: Run LSTM classifier
    FillInputTensor(lstm_interpreter.get(), lstm_input, 0);
    auto final_output = RunInference(lstm_interpreter.get());

    std::cout << "Violence Probability: " << final_output[0] << std::endl;
    if (final_output[0] > 0.5) {
        std::cout << "Violence Detected!" << std::endl;
    } else {
        std::cout << "No Violence Detected." << std::endl;
    }

    return 0;
}
